protocol/morse/sanctioned_endpoints_store.go

```diff
@@ -0,0 +1,190 @@
+			case protocolobservations.MorseSanctionType_MORSE_SANCTION_SESSION:
+				logger.Info().Msg("Adding session sanction for endpoint")
+
+				ses.addSessionSanction(
+					protocol.EndpointAddr(endpointObservation.GetEndpointAddr()),
+					endpointObservation.GetAppAddress(),
+					endpointObservation.GetSessionKey(),
+					sanctionData,
+				)
+			default:
+				logger.Warn().Msg("sanction type not supported by the store. skipping.")
+			}
+		}
+	}
+}
+
+// FilterSanctionedEndpoints removes sanctioned endpoints from the provided list.
+// It returns a filtered list containing only endpoints that don't have active sanctions.
+// This method is used during endpoint selection to avoid selecting endpoints that have been sanctioned.
+func (ses *sanctionedEndpointsStore) FilterSanctionedEndpoints(
+	allEndpoints []endpoint,
+	appAddr string,
+	sessionKey string,
+) []endpoint {
+	var filteredEndpoints []endpoint
+
+	for _, endpoint := range allEndpoints {
+		sanctioned, reason := ses.isSanctioned(endpoint.Addr(), appAddr, sessionKey)
+		if sanctioned {
+			logger := loggerWithEndpoint(ses.logger, appAddr, sessionKey, endpoint.Addr(), reason)
+			logger.Debug().Msg("Filtering out sanctioned endpoint")
+			continue
+		}
+		filteredEndpoints = append(filteredEndpoints, endpoint)
+	}
+
+	return filteredEndpoints
+}
+
+// addPermanentSanction adds a permanent sanction for an endpoint.
+// Permanent sanctions never expire and require manual intervention to remove.
+// These are used for serious errors like validation failures that may indicate malicious behavior.
+func (ses *sanctionedEndpointsStore) addPermanentSanction(
+	endpointAddr protocol.EndpointAddr,
+	sanctionData sanction,
+) {
+	ses.mu.Lock()
+	ses.permanentSanctions[endpointAddr] = sanctionData
+	ses.mu.Unlock()
+}
+
+// addSessionSanction adds a session-based sanction for an endpoint.
+// Session-based sanctions automatically expire after defaultSessionSanctionExpiration.
+// These are used for temporary issues like timeouts or connection problems.
+func (ses *sanctionedEndpointsStore) addSessionSanction(
+	endpointAddr protocol.EndpointAddr,
+	appAddr string,
+	sessionKey string,
+	sanctionData sanction,
+) {
+	// TODO_MVP(@adshmh): validate the session key format. Should not create a sanction entry if the session key is not valid.
+	key := sessionSanctionKey(appAddr, sessionKey, string(endpointAddr))
+	ses.sessionSanctions.Set(key, sanctionData, defaultSessionSanctionExpiration)
+}
+
+// isSanctioned checks if an endpoint has an active sanction.
+// It checks both permanent and session-based sanctions.
+func (ses *sanctionedEndpointsStore) isSanctioned(
+	endpointAddr protocol.EndpointAddr,
+	appAddr string,
+	sessionKey string,
+) (bool, string) {
+	// Check permanent sanctions first - these apply regardless of session
+	ses.mu.RLock()
+	sanctionRecord, hasPermanentSanction := ses.permanentSanctions[endpointAddr]
+	ses.mu.RUnlock()
```

Comment:
Wdyt of this instead?

```suggestion
	defer ses.mu.RUnlock()
	sanctionRecord, hasPermanentSanction := ses.permanentSanctions[endpointAddr]
```